### DISCUSSION TITLE
optimizer for prerendering

### DISCUSSION THREAD

**Discussion by RamIdeas** [parcel-bundler/parcel#5407] [Q&A]
I'm trying to write an optimizer for prerendering

I have the raw-url packager in use, [as resolved here](https://github.com/parcel-bundler/parcel/discussions/5393), but it seems the output from the packager hasn't been committed to the bundleGraph before calling my optimizer's `async optimize()`.

I was trying to create a http server that would serve the assets as needed from the in-memory bundleGraph, as below:

```js
const serveAssets = (bundleGraph) => {
	const bundles = bundleGraph.getBundles();

	const server = http.createServer(async (req, res) => {
		console.log('INCOMING', req.url);

		const filename = req.url.slice(1);
		const bundle = bundles.find((b) => b.name === filename);

		if (!bundle) {
			res.statusCode = 404;
			res.end();
			return;
		}

		const contents = await bundle.getMainEntry().getCode();
		res.end(contents);
	});

	return new Promise((resolve) => {
		server.listen(0, () => resolve(server));
	});
};
```

But the contents sent to `res.end(contents)` still has 32-char hex strings instead of the final filenames which the raw-url packager should (and does) replace.

Am I doing anything wrong? Is this even the correct strategy to implement prerendering?

### TL;DR

In an optimizer's `async optimize({ contents, bundle })` hook, `contents !== await bundle.getMainEntry().getCode()`. Is this by design?

**Comment by mischnic** [comment]
What does an optimizer have to do with an HTTP server? That should be in an reporter, e.g. https://github.com/parcel-bundler/parcel/blob/v2/packages/reporters/dev-server/src/ServerReporter.js.

> contents !== await bundle.getMainEntry().getCode()

Well, the raw url packager usually changes the assets contents: https://github.com/parcel-bundler/parcel/blob/635c973e3e584f7feb99f39c9e1963775e1f2ab1/packages/packagers/raw-url/src/RawUrlPackager.js#L15-L18

**Reply by RamIdeas** [reply]
Can a reporter modify assets?

As part of prerendering, I am serving the assets from an HTTP server to load them in a browser (puppeteer) and then update the html files with the stringified html from the DOM after all scripts have run.

I would like to do this within parcel because the html files are not entry files (they will have a hash in their filename) and I'd like to update their contents and have their filename hashes update to

> Well, the raw url packager usually changes the assets contents:

And packagers run before optimizers, right, so shouldn't their output be in the bundleGraph by the time the optimizers run?

**Reply by mischnic** [reply]
I see. Yes, pre-rendering would make sense in an optimizer.

> And packagers run before optimizers, right, so shouldn't their output be in the bundleGraph by the time the optimizers run?

The only thing the url packager does is rewriting the dependency hex-strings into paths. The filenames in these paths however still contain the hash placeholders and not the real filenames: `/index.HASH_REF_2e4c53d6217fd5a1e9ca638c3675f06e.js`. Why? Becauase, as you've said:

> I'd like to update their contents and have their filename hashes update to

1. Optimizers can change the output (and the hash), so optimizer doesn't know the name of its own bundle or of other bundles.
2. Furthermore, this all happens in parallel, so the other referenced bundle (e.g. a JS bundle) might not even be ready when you try to prerender it (the HTML bundle) .

<br>

I'm not sure if this is possible with the current architecture. There would have to be another optimize step that doesn't influence hashing and runs after all other bundles are ready.

**Reply by RamIdeas** [reply]
Regarding (1), I don't need to know the final filename/hash, I just need the current bundle name in the bundleGraph. It just needs to be consistent. It can, and should, change later if any of the assets change during the optimize step.

Regarding (2), I was operating under the assumption that optimizers run after all packagers have run.

Regarding your last point, that's what [parcel-plugin-prerender](https://www.npmjs.com/package/parcel-plugin-prerender) did in v1 (listened for the `.on('buildEnd')` event), but that isn't ideal because we _do, in fact,_ want to influence file hashes.

But I do agree an extra stage after `async optimize()` -- that allows for cross-bundle optimisations -- would be necessary to enable prerendering fully within the parcel pipeline.

**Reply by mischnic** [reply]
>  I don't need to know the final filename/hash, I just need the current bundle name in the bundleGraph. It just needs to be consistent

I assume your plan was to have the optimizer run on html bundles? In that case executing a referenced JS bundle isn't possible.

The way (we know of) at the moment is running the optimizer on the JS bundle, and changing the type with  `bundle.type = 'html';`. Because you probably also want a clientside JS bundle, you'd need a transformer that returns an additional JS asset (because the first one will become the HTML file).

> I was operating under the assumption that optimizers run after all packagers have run.

No, for every bundle (package + optimize) is scheduled. So one bundle might be already optimized while another is still being packaged.

> another optimize step that doesn't influence hashing and runs after all other bundles are ready.

Yeah, maybe there can be a system where this is factored into the hash...
